File: Code/0202.py
#!/usr/bin/python3

import copy

data = ''

lines = ''
with open(f'../Input/02.txt', 'r') as file:
    lines = file.readlines()

new_lines = []
for line in lines:
    new_lines.append(line[:-1])
lines = new_lines

def checkForJump(line):
    number = line[0]
    original = copy.deepcopy(line)
    result = True
    problem_index = 0

    first_result = True
    second_result = True

    for index, element in enumerate(line):
        if index == 0:
            continue
        difference = int(number) - int(line[index])
        number = line[index]

        if difference > 0:
            problem_index = index - 1
            result = False
            break

    if result == False:
        del line[problem_index]

        number = line[0]

        result = True

        # Test it again
        for index, element in enumerate(line):
            if index == 0:
                continue
            difference = abs(int(number) - int(line[index]))
            number = line[index]

            if difference > 3 or difference <= 0:
                first_result = False
                break
    else:
        first_result = True




    # Again!
    line = original
    problem_index = 0
    line.reverse()
    number = line[0]

    for index, element in enumerate(line):
        if index == 0:
            continue
        difference = int(number) - int(line[index])
        number = line[index]

        if difference > 0:
            problem_index = index
            result = False
            break

    if result == False:
        del line[problem_index]

        result = True
        number = line[0]

        # Test it again
        for index, element in enumerate(line):
            if index == 0:
                continue
            difference = abs(int(number) - int(line[index]))
            number = line[index]

            if difference > 3 or difference <= 0:
                second_result = False
                break
    else:
        second_result = True

    if first_result or second_result:
        return True
    else:
        return False



def checkForConsistency(line):
    number = line[0]
    original = copy.deepcopy(line)
    result = True
    problem_index = 0
    first_result = True
    second_result = True

    for index, element in enumerate(line):
        if index == 0:
            continue
        if int(line[index]) > int(number):
            number = int(line[index])
        else:
            problem_index = index - 1
            result = False
            break

    if not result:
        del line[problem_index]

        result = True

        number = line[0]

        for index, element in enumerate(line):
            if index == 0:
                continue
            if int(line[index]) > int(number):
                number = int(line[index])
            else:
                first_result = False
    else:
        first_result = True

    # Check the other direction
    problem_index = 0
    line = original
    number = int(line[0])

    for index, element in enumerate(line):
        if index == 0:
            continue
        if int(line[index]) < int(number):
            number = int(line[index])
        else:
            problem_index = index - 1
            result = False
            break

    if not result:
        del line[problem_index]

        result = True

        number = line[0]

        for index, element in enumerate(line):
            if index == 0:
                continue
            print(line[index], number)
            if int(line[index]) < int(number):
                number = int(line[index])
            else:
                second_result = False
    else:
        second_result = True

    if first_result or second_result:
        return True
    else:
        return False



safe_reports = 0

for index, element in enumerate(lines):
    if index != 1:
        continue
    first = checkForJump(lines[index].split())
    second = checkForConsistency(lines[index].split())

    print(first, second)

    if first and second:
        safe_reports += 1

print('-------')
print(safe_reports)

File: Code/0202.lua
#!/usr/bin/lua

input = '../Input/02.txt'

-- Read the file into "data"
file = {}
for line in io.lines(input) do
    file[#file + 1] = line
end

-- Split on whitespace
lines = {}
for i=1, #file do
    table.insert(lines, {})
    count = 1
    for substring in file[i]:gmatch("%S+") do
       lines[i][count] = substring
       count = count + 1
    end
end


function checkForJump(line)
    number = line[1]
    problems = 0

    for j=2, #line do
        difference = math.abs(number - line[j])
        number = line[j]

        if difference > 3 or difference <= 0 then
            problems = problems + 1
        end
    end

    print('problems', problems)

    if problems > 1 then
        return false
    elseif problems <= 1 then
        return true
    end
end


function checkForConsistency(line)
    number = line[1]
    result = true
    first_problems = 0
    for i=2, #line do
        if tonumber(line[i]) > tonumber(number) then
            number = tonumber(line[i])
        else
            first_problems = first_problems + 1
        end
    end


    number = tonumber(line[1])
    second_problems = 0
    for i=2, #line do
        if tonumber(line[i]) < tonumber(number) then
            number = tonumber(line[i])
        else
            second_problems = second_problems + 1
        end
    end

    if first_problems <= 1 or second_problems <= 1 then
        return true
    else
        return false
    end
end



safe_reports = 0

for i=1, #lines do
    if i==3 then break end
    first = checkForJump(lines[i])
    second = checkForConsistency(lines[i])

    print(first, second)

    if first and second then
        safe_reports = safe_reports + 1
    end
end

print(safe_reports)
